src/app/shared/services/doctors.service.ts

```diff
@@ -1,32 +1,160 @@
+  private visibleDoctorsSubject = new BehaviorSubject<Doctor[]>([]);
+  private visibleSpecialtiesSubject = new BehaviorSubject<DoctorsSpecialty[]>(
+    []
+  );
+
+  private topDoctorsSubject = new BehaviorSubject<Doctor[]>([]);
+  private paginationSizeSubject = new BehaviorSubject<number>(0);
+
+  private paginationCurrentSubject = new BehaviorSubject<number>(0);
+  private sortingTypeSubject = new BehaviorSubject<string>('');
+  private searchQuerySubject = new BehaviorSubject<string>('');
+
+  private subsription!: Subscription;
+
+  visibleDoctors = this.visibleDoctorsSubject.asObservable();
+  visibleSpecialties = this.visibleSpecialtiesSubject.asObservable();
+
+  paginationCurrent = this.paginationCurrentSubject.asObservable();
+  paginationSize = this.paginationSizeSubject.asObservable();
+  sortingType = this.sortingTypeSubject.asObservable();
+
+  topDoctors = this.topDoctorsSubject.asObservable();
+
+  private firebaseDoctors: Doctor[];
+  private firebaseSpecialties: Object;
 
   constructor(private db: AngularFireDatabase) {
-    this.doctors = this.db.list('/doctors').valueChanges() as Observable<
-      Doctor[]
-    >;
+    this.firebaseDoctors = [];
+    this.firebaseSpecialties = {};
   }
 
-  getTopDoctors(): Observable<Doctor[]> {
-    return this.doctors.pipe(
-      map((doctors) => doctors.filter((doctor) => doctor.top))
-    );
+  ngOnDestroy(): void {
+    this.subsription.unsubscribe();
+  }
+
+  loadData(): void {
+    this.subsription = this.db
+      .list('/doctors')
+      .valueChanges()
+      .subscribe((data) => {
+        this.firebaseDoctors = data as Doctor[];
+        this.updateVisibleDoctors();
+        this.calculateTopDoctors();
+      });
+
+    this.loadSpecialties();
   }
 
-  getSpecialties(): Observable<string[]> {
-    return this.doctors.pipe(
-      map((doctors) =>
-        Array.from(new Set(doctors.map((doctor) => doctor.specialization)))
-      )
+  loadSpecialties(): void {
+    this.db.database.ref('specialties').once('value', (snap) => {
+      this.firebaseSpecialties = snap.val();
+      const specialties = Object.entries(snap.val()).map(([id, name]) => ({
+        id,
+        name: name as string,
+        checked: false,
+      }));
+      this.visibleSpecialtiesSubject.next(specialties);
+    });
+  }
+
+  private updateVisibleDoctors(): void {
+    const specialties = this.visibleSpecialtiesSubject.value
+      .filter((v) => v.checked)
+      .map((v) => v.id);
+
+    const hasSearchFilter = !!this.searchQuerySubject.value;
+    const searchLowerStr = this.searchQuerySubject.value?.toLowerCase() || '';
+    const hasSpecialtiesFilter = !!specialties && specialties.length > 0;
+
+    const filteredDoctors = this.firebaseDoctors.filter((doc) => {
+      if (hasSearchFilter && !doc.name.toLowerCase().includes(searchLowerStr)) {
+        return false;
+      }
+      if (hasSpecialtiesFilter && !specialties?.includes(doc.specialization)) {
+        return false;
+      }
```

Comment:
The if statements for filtering doc objects can be combined for better readability and performance.

```suggestion
      if (hasSearchFilter && !doc.name.toLowerCase().includes(searchLowerStr)
        || hasSpecialtiesFilter && !specialties?.includes(doc.specialization)) {
        return false;
      }
```

Reply:
Well, yes, perhaps it will be better this way.

Reply:
I looked over it again and had even better idea:

```suggestion
      .filter((doc) => {
        return !((hasSearchFilter && !doc.name.toLowerCase().includes(searchLowerStr)) || 
        (hasSpecialtiesFilter && !specialties?.includes(doc.specialization))
  );
});
```

Reply:
Yes, looks pretty good!

Reply:
Actually, I think last code will have worse performance because you need to calculate both parts of || statement. There are a lot of cases where calculating ```(hasSearchFilter && !doc.name.toLowerCase().includes(searchLowerStr)``` is enought to make a decision to remove element from array.
Let's keep the original code, it have enough readability and performance

Reply:
This isn't true. The logical OR (||) is lazy, if the first part is true it would not proceed to check the next parts. So, performance wise this is how it works.
```js
function a() {
	console.log('a')
	return true
}

function b() {
	console.log('b')
	return true
}

console.log(a() || b()) // Output: "a" -> true
```

Reply:
Anyway, let’s keep original code, it has more readability

Reply:
Probably you right, I’ve read it wrong) 
!((condition && condition) || (condition && condition)) is hard to read for me 🙃